### ISSUE TITLE
Verify TMC Integration works for registering TCE clusters

### ISSUE THREAD

**Issue by joshrosso** [vmware-tanzu/community-edition#2992]

**Comment by joshrosso** [comment]
@berndtj are you and/or someone from your team able to look at this proposal and let us know if it makes sense based on  your understanding of needs for TCE+TMC interop?

**Comment by mfine30** [comment]
Overall this looks like what I would expect, at least from a product perspective. I have a few questions just to confirm things though.

I see an Issue titled: "TMC must be able to use the tanzu-framework library to create TCE-specific workload clusters" but don't understand enough of the actual issue under it #2716 to know if that will solve the tile description. I assume it will since you linked them though?

Will TMC also be able to understand that the workload clusters are TCE workload clusters or is it just the management clusters? I'd vote for both if possible.

**Comment by joshrosso** [comment]
@mfine30 thanks for the follow-up.

> I see an Issue titled: "TMC must be able to use the tanzu-framework library to create TCE-specific workload clusters" but don't understand enough of the actual issue under it https://github.com/vmware-tanzu/community-edition/issues/2716 to know if that will solve the tile description. I assume it will since you linked them though?

Yes, https://github.com/vmware-tanzu/community-edition/issues/2716 aims to solve a case where the tanzu `management-cluster` and `cluster` CLI plugins (from tanzu-framework) can operate against TCE and TKG clusters with the same binary.

I'd need guidance from TMC if more than this is needed.

> Will TMC also be able to understand that the workload clusters are TCE workload clusters or is it just the management clusters? I'd vote for both if possible.

Based on the current plumbing, TMC will know that the workload clusters came from TMC by nature of being registered to t he management cluster that made them. However, it will not be obvious in the same way by looking **only** at the workload cluster.

Can ya'll expand on whether this is a requirement or more-so a nice to have?

**Comment by berndtj** [comment]
I think what @mfine30 is getting at is that ideally a workload cluster attached only (i.e. registered management cluster) should also be detectable by TMC as coming from TCE (or TKG for that matter).

cc: @steven-zou

**Comment by mfine30** [comment]
Here are the user stories that I have in mind driving this:
1. As a TMC/TCE user, I can register a TCE management cluster to TMC
2. As a TMC/TCE user with a registered TCE management cluster, I can use that management cluster to provision a TCE workload cluster
3. As a TMC/TCE user, I can identify that a management cluster registered to TMC is TCE vs TKG
4.  As a TMC/TCE user, I can identify that a TCE workload cluster provisioned by TMC/TCE is a TCE vs TKG
5. As a TMC/TCE user, with a TCE workload cluster that I created outside of TMC (and the management cluster is not registered to TMC), when I attach the workload cluster to TMC I can tell that it is a TCE workload cluster

**Comment by joshrosso** [comment]
@berndtj @mfine30, makes sense. 

Can you please verify `4` and `5` **are** required out of the gate?

**Comment by mfine30** [comment]
Yes I'd consider 4 and 5 as required as well. I worry that only doing 1-3 creates a sub-par UX that looks to customers like we forgot to finish it.

**Comment by joshrosso** [comment]
k, i'll spec it out in the proposal above.

@mfine30 and @berndtj how do you do this detection for `4` and `5` with TKG **workload** clusters today?

**Comment by mfine30** [comment]
@joshrosso – thanks for pushing on that and making us think through it. From chatting with @berndtj, I realized that I thought we did more in TMC than TMC does today. TMC currently only shows the cloud vs the provider for an attached cluster (i.e. it's a vSphere cluster for an attached cluster). 

I think that means for this track of work:
* item `4` can be inferred by TMC based on the management cluster rather than introspecting the workload cluster; @berndtj you ok with that approach?
* item `5` should not be considered a requirement

**Comment by joshrosso** [comment]
perfect, that's exactly what I was hoping ya'll would say.

Make sense and I do agree it'd be a solid future-term thing to expose.

@berndtj can you throw us a :+1: you're on the same page with what @mfine30 said?

**Comment by steven-zou** [comment]
From the TMC LCM engineering perspective, we'd like the current TKG LCM extension can also support managing the LCM of TCE clusters to reduce efforts and code complexities. To achieve this goal, there are some needs that need to be met. 

The first and top thing is the **PRIMARY NEED2** of this proposal:
>TMC can use the same tanzu CLI to interact with TCE management clusters

It should be noted here that the current TKG LCM extension is referring to the client library in the mirror tanzu framework gitlab repository `core-build/mirrors_github_tanzu-framework` instead of the github tanzu framework `github.com/vmware-tanzu/tanzu-framework`. I'm not sure if there are incompatibility issues between these two repos when supporting TCE. We have to think about it ahead.

Then some metadata for detecting and identifying TKG/TCE clusters, might include:

- `metadata.annotations[TKGVERSION]` : for getting TKG/TCE version
- `metadata.labels[tanzuKubernetesRelease]`: for upgrading
- `metadata.annotations[edition]`: for distinguishing TKG/TCE  (so far, hardcoded edition related info to `TKG` at TMC side)

cc @berndtj  and @renmaosheng

**Comment by steven-zou** [comment]
Additionally, story `item 5` is really not part of LCM stories.

**Comment by joshrosso** [comment]
> It should be noted here that the current TKG LCM extension is referring to the client library in the mirror tanzu framework gitlab repository core-build/mirrors_github_tanzu-framework instead of the github tanzu framework github.com/vmware-tanzu/tanzu-framework. I'm not sure if there are incompatibility issues between these two repos when supporting TCE. We have to think about it ahead.

Who can figure this compatibility question out?

As part of this work, we'll ensure the upstream libs (tanzu-framework) have this capability. Who will own the reconciliation of the mirror/downstream dep?

**Comment by renmaosheng** [comment]
@Maharshi Bhatt is the major contact from TKG downstream side, we can confirm with Maharshi on the compatibility questions.  to Echo Steven Zou's point, we want the TKG/TCE release can share the same tanzu-framework sha, so that TMC doesn't need to treat TKG/TCE separately. we are also working with TKG to define an API instead of using the client library code, we want to align with TCE as well to use the same API in a long-term perspective.

**Comment by joshrosso** [comment]
@renmaosheng 

> we want the TKG/TCE release can share the same tanzu-framework sha, so that TMC doesn't need to treat TKG/TCE separately.

Unfortunately, this will not be possible. Tanzu-framework should provide backwards compatibility guarantees. We cannot pin the community project on versions of our downstream product.

**Comment by joshrosso** [comment]
Moving this out to re-evaluate in v0.12.0.

We've decided to take a slightly different approach to ensure we can ship v0.11.0.

That work is tracked here: https://github.com/vmware-tanzu/community-edition/issues/3285
